mito-ds/monorepo
# Ipywidgets 8.0

# Description

We've observed users installing Mito on Python 3.10+ run into the same issues over and over. 

```
$ python -m mitoinstaller install --no-cache-dir
Starting install...
Upgrade mitoinstaller
Setting up environment
Check dependencies
Remove mitosheet3 if present
Install mitosheet
Activate extension
Traceback (most recent call last):
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/mitoinstaller/installer_steps/installer_step.py", line 44, in execute
    self.execution_function()
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/mitoinstaller/installer_steps/mitosheet_installer_steps.py", line 77, in install_step_mitosheet_activate_notebook_extension
    run_command([sys.executable, "-m", 'jupyter', 'nbextension', 'install', '--py', '--user', 'mitosheet'])
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/mitoinstaller/commands.py", line 50, in run_command
    raise Exception(get_output(completed_process))
Exception: Traceback (most recent call last):
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/bin/jupyter-nbextension", line 8, in <module>
    sys.exit(main())
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/jupyter_core/application.py", line 269, in launch_instance
    return super().launch_instance(argv=argv, **kwargs)
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/traitlets/config/application.py", line 976, in launch_instance
    app.start()
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/nbclassic/nbextensions.py", line 972, in start
    super().start()
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/jupyter_core/application.py", line 258, in start
    self.subapp.start()
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/nbclassic/nbextensions.py", line 702, in start
    self.install_extensions()
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/nbclassic/nbextensions.py", line 675, in install_extensions
    full_dests = install(self.extra_args[0],
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/nbclassic/nbextensions.py", line 203, in install_nbextension_python
    m, nbexts = _get_nbextension_metadata(module)
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/nbclassic/nbextensions.py", line 1107, in _get_nbextension_metadata
    m = import_item(module)
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/traitlets/utils/importstring.py", line 38, in import_item
    return __import__(parts[0])
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/mitosheet/__init__.py", line 31, in <module>
    from mitosheet.mito_widget import MitoWidget, sheet
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/mitosheet/mito_widget.py", line 17, in <module>
    from ipywidgets import DOMWidget
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/ipywidgets/__init__.py", line 25, in <module>
    from .widgets import *
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/ipywidgets/widgets/__init__.py", line 20, in <module>
    from .widget_selection import RadioButtons, ToggleButtons, ToggleButtonsStyle, Dropdown, Select, SelectionSlider, SelectMultiple, SelectionRangeSlider
  File "/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/site-packages/ipywidgets/widgets/widget_selection.py", line 9, in <module>
    from collections import Mapping, Iterable
ImportError: cannot import name 'Mapping' from 'collections' (/Users/aarondiamond-reivich/opt/anaconda3/envs/mitoenv/lib/python3.10/collections/__init__.py)

```
This error is thrown in the ipywidgets codebase. The error is caused because in Python 3.10, instead of importing `from collections import Mapping`, you must import as `from collections.abc import Mapping`.

Ipywdigets 8.0 resolves this import issue. See [here](https://github.com/jupyter-widgets/ipywidgets/blob/3dbd281a7488da6d2e2fc10d14df58041963eaa6/python/ipywidgets/ipywidgets/widgets/widget_selection.py).

This PR updates Mito to work with ipywidgets 8.0

See investigation and documentation about this upgrade process [here](https://www.notion.so/trymito/Python-3-10-Error-cbec7bf27bae42e58083ece3bc052082)

In upgrading my Python version to 3.10, I also ran into issues with the development environmnent. Specifically:
1. `ensure_python` is deprecated on 3.10, so I switched to `python_requires`
2. removing `ensure_python` caused GitHub tests to not pass. I was able to resolve this by following the error message and updating `InitializeParameters` to `IInitializeParameters` in `src/jupyter/widget.tsx`

Fixing 2 above, caused Mito on JupyterLab 2 to not work, however. It says:
```
"src/jupyter/widget.tsx(69,39): error TS2724: '"/Users/aarondiamond-reivich/Documents/saga/monorepo/mitosheet/node_modules/@jupyter-widgets/base/lib/widget".WidgetView' has no exported member named 'IInitializeParameters'. Did you mean 'InitializeParameters'?"
```

# Testing

To test:
1. Install python 3.10
2. Checkout the main branch
4. Remove the ensure_python import and use in the setup.py file
5. Create a new virtual evironment and run the commands to start Mito (note you might need to install a xcode upgrade, just read the error message)
6. Notice that that you receive the error ` ImportError("cannot import name 'Mapping' from 'collections' `
7. Checkout this branch
8. Create a new virtual evironment and run the commands to start Mito
9. See that Mito works 

# Documentation

Note if any new documentation needs to addressed or reviewed.
